advanced_functionality/pytorch_extending_our_containers/pytorch_extending_our_containers.ipynb

```diff
@@ -0,0 +1,701 @@
+   "metadata": {},
+   "source": [
+    "## SageMaker Python SDK Local Training\n",
+    "To represent our training, we use the Estimator class, which needs to be configured in five steps. \n",
+    "1. IAM role - our AWS execution role\n",
+    "2. train_instance_count - number of instances to use for training.\n",
+    "3. train_instance_type - type of instance to use for training. For training locally, we specify `local`.\n",
+    "4. image_name - our custom PyTorch Docker image we created.\n",
+    "5. hyperparameters - hyperparameters we want to pass.\n",
+    "\n",
+    "Let's start with setting up our IAM role. We make use of a helper function within the Python SDK. This function throw an exception if run outside of a SageMaker notebook instance, as it gets metadata from the notebook instance. If running outside, you must provide an IAM role with proper access stated above in [Permissions](#Permissions)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sagemaker import get_execution_role\n",
+    "\n",
+    "role = get_execution_role()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Fit, Deploy, Predict\n",
+    "\n",
+    "Now that the rest of our estimator is configured, we can call `fit()` with the path to our local CIFAR10 dataset prefixed with `file://`. This invokes our PyTorch container with 'train' and passes in our hyperparameters and other metadata as json files in /opt/ml/input/config within the container to our program entry point defined in the Dockerfile.\n",
+    "\n",
+    "After our training has succeeded, our training algorithm outputs our trained model within the /opt/ml/model directory, which is used to handle predictions.\n",
+    "\n",
+    "We can then call `deploy()` with an instance_count and instance_type, which is 1 and `local`. This invokes our PyTorch container with 'serve', which setups our container to handle prediction requests as defined [here](https://github.com/aws/sagemaker-pytorch-container/blob/master/src/sagemaker_pytorch_container/serving.py#L103). What is returned is a predictor, which is used to make inferences against our trained model.\n",
+    "\n",
+    "After our prediction, we can delete our endpoint.\n",
+    "\n",
+    "We recommend testing and training your training algorithm locally first, as it provides quicker iterations and better debuggability."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Lets set up our SageMaker notebook instance for local mode.\n",
+    "!/bin/bash ./utils/setup.sh"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sagemaker.estimator import Estimator\n",
+    "\n",
+    "hyperparameters = {'epochs': 1}\n",
+    "\n",
+    "instance_type = 'local'\n",
```

Comment:
Since this is CIFAR, I think we should have logic to use the GPU if one is available.  e.g.

```python
instance_type = 'local'

if subprocess.call('nvidia-smi') == 0:
    ## Set type to GPU if one is present
    instance_type = 'local_gpu'
```

Reply:
Added!